packages/fxa-auth-server/lib/routes/emails.js

```diff
@@ -367,157 +358,143 @@ module.exports = (
+        // the target email address before we can check it with customs.
+        await customs.check(request, account.email, 'recoveryEmailVerifyCode');
+        // Check if param `type` is specified and equal to `secondary`
+        // If so, verify the secondary email and respond
+        if (type && type === 'secondary') {
+          await verifySecondaryEmail(account);
+          return {};
+        }
 
-                // Attempt to verify email token not associated with account
-                if (!isEmailVerification) {
-                  throw error.invalidVerificationCode();
-                }
-
-                // User is attempting to verify a secondary email that has already been verified.
-                // Silently succeed and don't send post verification email.
-                if (matchedEmail.isVerified) {
-                  log.info('account.verifyEmail.secondary.already-verified', {
-                    uid,
-                    code,
-                  });
-                  return P.resolve();
-                }
-
-                return db.verifyEmail(account, code).then(() => {
-                  log.info('account.verifyEmail.secondary.confirmed', {
-                    uid,
-                    code,
-                  });
-
-                  return mailer.sendPostVerifySecondaryEmail([], account, {
-                    acceptLanguage: request.app.acceptLanguage,
-                    secondaryEmail: matchedEmail.email,
-                    service,
-                    uid,
-                  });
-                });
+        const isAccountVerification = butil.buffersAreEqual(
+          code,
+          account.emailCode
+        );
+
+        let device;
+
+        try {
+          device = await db.deviceFromTokenVerificationId(uid, code);
+        } catch (err) {
+          if (err.errno !== error.ERRNO.DEVICE_UNKNOWN) {
+            log.error('Account.RecoveryEmailVerify', {
+              err,
+              uid,
+              code,
+            });
+          }
+        }
+
+        await accountAndTokenVerification(isAccountVerification, account);
+
+        if (device) {
+          const devices = request.app.devices;
+          const otherDevices = devices.filter(d => d.id !== device.id);
+          await push.notifyDeviceConnected(uid, otherDevices, device.name);
+        }
+
+        // If the account is already verified, the link may have been
+        // for sign-in confirmation or they may have been clicking a
+        // stale link. Silently succeed.
+        if (account.emailVerified) {
+          return {};
+        }
+
+        // Any matching code verifies the account
+        await signupUtils.verifyAccount(request, account, request.payload);
+
+        return {};
+
+        async function verifySecondaryEmail(account) {
+          let matchedEmail;
+          const emails = await db.accountEmails(uid);
+          const isEmailVerification = emails.some(email => {
```

Comment:
What happens if someone programatically calls this endpoint with a primary email and specifies a `type` of `secondary`? Do we handle that correctly? And what about the reverse, they call the endpoint with a secondary email and specify a type of `primary`?

Reply:
The short answer is I don't know.  This was the `then` handler from https://github.com/mozilla/fxa/pull/2324/files#diff-a8b1e870a711340cf536638ff5ec0730L381.  

For this patch I really did try to just collapse the promise chains and use async/await.  I can look into those scenarios, but I don't think those type of changes (if any needed) should go into this patch.

Reply:
Ok, so the endpoint doesn't accept an email, and the only accepted value for type is 'secondary' (not in validation but in the `if`).  I think the question is: what happens if someone with a code for an account's primary email set `type` to 'secondary' in the payload.  In that case, the email will be verified, but the user will received a different post-verify email, and https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/lib/routes/utils/signup.js#L16 get bypassed.